core/src/main/scala/za/co/absa/spline/harvester/dispatcher/httpdispatcher/rest/AuthenticationFactory.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2020 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.harvester.dispatcher.httpdispatcher.rest
+
+import java.time.{Duration, Instant}
+
+import scalaj.http.{Base64, HttpRequest}
+
+import scala.collection.mutable.Map
+import scala.collection.immutable.Map
+import org.apache.spark.internal.Logging
+
+trait Authentication {
+  def createRequest(httpRequest: HttpRequest,authentication: scala.collection.immutable.Map[String,String]):HttpRequest
+}
+
+case class NoAuthentication(authentication: scala.collection.immutable.Map[String, String]) extends Authentication{
+  override def createRequest(httpRequest: HttpRequest, authentication: scala.collection.immutable.Map[String, String]): HttpRequest = httpRequest
+}
+case class ClientCredentialsAuthentication(authentication: scala.collection.immutable.Map[String, String]) extends Authentication with Logging {
+  case class Token(tokenValue: String, expirationTime: Instant)
+  private val tokenCache: scala.collection.mutable.Map[String, Token] = scala.collection.mutable.Map.empty[String, Token]
+  private val clientId: String = authentication.getOrElse("clientId","")
+  private val clientSecret: String = authentication.getOrElse("clientSecret","")
+  private val tokenUrl: String = authentication.getOrElse("tokenUrl","")
+  private val grantType: String = authentication.getOrElse("grantType","")
+  private val scope: String = authentication.getOrElse("scope","")
+
+  private def isTokenExpired(expirationTime: Instant): Boolean = {
+    Instant.now().isAfter(expirationTime.minusSeconds(300) )
+  }
```

Comment:
Good idea to extract the expiration checking logic into a separated method. I would go even further and use a `isTokenValid` method where you can check both, that it is defined and the expiration.